### ISSUE TITLE
Balloon movement inconsistant

### ISSUE THREAD

**Issue by terracoda** [phetsims/balloons-and-static-electricity#128]
In my last interview, inconsistent balloon movement occurred several times. I did not notice this behaviour in other interviews on the same system.
In several instances, a single directional key press sent the balloon to the extreme edge of the Play Area. This happened first (Green Balloon) from the middle of the Play Area up to the top with one W press. 
Later, after 3 normal steps the Yellow Balloon reached the right edge of Sweater and then the next A key press sent it right across the Sweater in one go. A subsequent D key press sent the Balloon all the way to the Wall and then another A key press sent it all the way back to left of Play Area and then another D key press back over to Wall. 
Then there was some normal behaviour and a little later the Green Balloon was stepping towards the Sweater and when it got to the right side of the Sweater, the next A key press sent it left across the Sweater in one go.
The Participant was not holding the key down and not using the shift key.
The system details are: 
- Windows 10, 
- JAWS 17 demo mode
- Firefox 43.0.1
- Lenova LapTop T520, core i7

@jessegreenberg have you seen this kind of thing before?

**Comment by jessegreenberg** [comment]
I have observed this issue a couple of times though without consistency and only on JAWS + IE.  Thank you for the detailed report and the system information!

This is something we will need to investigate.  I am not sure if this is related to sim performance or the way that JAWS handles input events.  When I observed the issue, the mouse dragging performed well, so I am leaning toward JAWS event handling.  There seems to be a delay in the keyup/down events.  I will track the events with JAWS and see if we can pinpoint the delay.

**Comment by terracoda** [comment]
@jessegreenberg, good to know you had seen this before. On the Surface Pro 3 (Windows 10 and IE11), there seemed to be some "delayed" movement. For example, user made a couple of key presses that did not move the balloon and then it made a big move like it was playing catch up.  But in the cases described above the user did not make multiple key presses.

**Comment by jessegreenberg** [comment]
Ok, thanks @terracoda.  Will be looking at this today.

**Comment by jessegreenberg** [comment]
BalloonNode instruments a combination of ```keyup``` and ```keydown``` event listeners.  Since only one key event is fired at a time, we use this implementation to track the 'state' of the keyboard by updating a map of which keys are down.  This also lets us avoid the HTML5 'press and hold' delay since we update the model depending on the state of the keyboard map in milliseconds.

When no AT is being used and focus is on the balloon, we observe the following events for 'press and hold':
![hold-event](https://cloud.githubusercontent.com/assets/6396244/13086135/06a9fc02-d49f-11e5-8319-0145d8b7cd21.PNG)


And the following events when tapping keys:
![tapped events](https://cloud.githubusercontent.com/assets/6396244/13086105/e3acfc90-d49e-11e5-9673-7b721aac820f.PNG)

We should see the same event order when the AT is being used.

**Comment by jessegreenberg** [comment]
With NVDA and Firefox:

Press and hold:
![firefox](https://cloud.githubusercontent.com/assets/6396244/13086335/0167fcf2-d4a0-11e5-8b26-c20b819a4a33.PNG)

Tappinig:
![tapping-firefox](https://cloud.githubusercontent.com/assets/6396244/13086342/060110aa-d4a0-11e5-887e-9ca179f46dc0.PNG)

Looks good with NVDA + Firefox :+1:

**Comment by jessegreenberg** [comment]
Firefox started to behave extremely poorly while JAWS turned on.  Reinstalling fixed issues.  Back to testing for key events.

**Comment by jessegreenberg** [comment]
Just tested with with Firefox 44 + JAWS 17, Windows 10.  Could not reproduce issues, same event pattern as https://github.com/phetsims/balloons-and-static-electricity/issues/128#issuecomment-184812698

**Comment by jessegreenberg** [comment]
Able to reproduce with JAWS 17 + IE 11, *and* IE 11 with the developer tools enabled.  This makes me think it is a performance issue.

When observing the keyboard events, the order is correct as above, but there is a delay in the ```keyup``` event, which means that the sim thinks the keys are pressed for longer than they actually are.

**Comment by jessegreenberg** [comment]
Now convinced this is a performance issue.  Problem is especially bad when JAWS + Dev tools enabled.  Less observable when JAWS disabled, but still present when dev tools enabled.  Not as observable with NVDA, likely because this reader is light weight and doesn't hog as much of the CPU.

We will need to consider scenery related performance optimizations to resolve this issue.

**Comment by jessegreenberg** [comment]
I am wondering if some strange balloon behavior could be related to the lack of keyup event support on certain devices? See
http://stackoverflow.com/questions/28944171/e-keycode-in-ios-safari-on-keyup-event

While playing with the sim on an iPad, i noticed that pressing W, A, S, or D with a bluetooth keyboard will send the balloon all the way to one end of the play area.  Perhaps this is what we are seeing on other platforms as well.

**Comment by jessegreenberg** [comment]
http://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.2.0-reader.2/balloons-and-static-electricity_en.html?reader&keyData

Above version has the reader output replaced with keyUp information to check if keyup is supported.

**Comment by jessegreenberg** [comment]
This issue is caused by #164, and the status will be tracked there.  Closing.
